### DISCUSSION TITLE
Structure to show the algorithm of division of polynomials

### DISCUSSION THREAD

**Discussion by psads-git** [typst/typst#3149] [Q&A] [answered]
I would to reproduce, using `typst`, the following that did by using `LaTeX`:

![image](https://github.com/typst/typst/assets/75945439/76438340-128b-47eb-a595-9e283c63beb0)

Could you please direct me? Thanks!

**Comment by xkevio** [comment, marked as answer]
You can probably cook something together with the `tablex` package here. I've made a basic example that you can customize:
```typ
#import "@preview/tablex:0.0.7": *

#tablex(columns: 5, auto-lines: false, stroke: 0.5pt, align: right,
  $x^3$, $0x^2$, $x$, $1$, vlinex(), $x + 1$,
  hlinex(start: 4, end: 5),
  $-x^3$, $-x^2$, [], [], $x^2 - x$,
  hlinex(end: 4),
  [], $-x^2$, $x$, $1$, [],
  [], $x^2$, $x$, [], [],
  hlinex(start: 1, end: 4),
  [], [], $2x$, $1$, [],
)
```

![image](https://github.com/typst/typst/assets/13004777/44a5caaa-62ea-4da4-8d58-a093581d5e2a)

**Reply by psads-git** [reply]
Thanks, @xkevio: that is exactly what I was looking for!

**Comment by PgBiel** [comment]
Here's an updated version using the new built-in grids and tables:
```typ
// MIT-0 licensed, feel free to use
#let longdiv(all-columns, ..cells) = {
  let longdiv = grid
  let cols = if type(all-columns) == array {
    all-columns.len()
  } else if type(all-columns) == int {
    all-columns
  } else {
    1
  }
  set grid(
    columns: cols,
    inset: 5pt,
    align: right,
    stroke: (x, y) => (
      // Add left stroke to the last column
      left: if x == cols - 1 { black },

      bottom: if (
        // Add bottom stroke to the top right cell
        y == 0 and x == cols - 1
  
        // Add bottom stroke every two rows (calc.odd check),
        // but for one less column each time
        or x < cols - 1 and calc.odd(y) and x + 1 >= y / 2
      ) {
        black
      }
    ),
  )
  grid(..cells)
}
```

Usage: specify number of columns (`#5` below - the hash is to ensure it is parsed as an integer and not text), and make sure you follow the expected structure. **Careful with the commas** (they are needed to skip columns).

```typ
$
longdiv(
  #5,
  x^10, x^4, x^2, x^4, x^5 + x^2,
  -x^2, -x^5, , , = x^5 + x^2,
  , x^2, x^50, x^10, ,
  , -x^3, -x^50, ,
  , , , x^50, x^10,
  , , , -x^50, -x^10,
  , , , , 234,
)
$
```
![image](https://github.com/user-attachments/assets/8a6778ed-69a5-49e7-98ae-95fba7e0aefa)

**Reply by psads-git** [reply]
Thanks, @PgBiel! That is a fantastic solution!

**Reply by bluss** [reply]
As of Typst 0.13, this example still works but it benefits from using `grid(..cells.pos().map(math.equation))`.   It doesn't show in the example, but a single term like `x` will show up upright and not italic without this change.

(and if you do that, probably pass `..cells.named()` separately, to allow more parameter overrides).

**Reply by psads-git** [reply]
Thanks, @bluss, for let us know that!
